### PR TITLE
[ENH][FIX]  Add Configurable Line Endings

### PR DESCRIPTION
**Pull request by DragonCrafted87**
Fix for Issue #224 

Add new Parameter
--line-ending

that controls how the line endings are set for the output file

### CODE REVIEW

File: mdformat/_util.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def is_md_equal(
     return html_texts["md1"] == html_texts["md2"]
 
 
-def atomic_write(path: Path, text: str) -> None:
+def atomic_write(path: Path, text: str, line_ending: Optional[str] = None) -> None:
```
Comment (hukkin) [inline mdformat/_util.py:81]:
```suggestion
def atomic_write(path: Path, text: str, line_ending: str) -> None:
```

File: mdformat/_util.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def atomic_write(path: Path, text: str) -> None:
     """
     fd, tmp_path = tempfile.mkstemp(dir=path.parent)
```
Comment (hukkin) [inline mdformat/_util.py:88]:
```suggestion
    newline = "\r\n" if line_ending == "crlf" else "\n"
    fd, tmp_path = tempfile.mkstemp(dir=path.parent)
```

File: mdformat/_util.py [hunk 3]
```diff
@@ -87,7 +87,7 @@ def atomic_write(path: Path, text: str) -> None:
...
     try:
-        with open(fd, "w", encoding="utf-8") as f:
+        with open(fd, "w", encoding="utf-8", newline=line_ending) as f:
```
Comment (hukkin) [inline mdformat/_util.py:90]:
```suggestion
        with open(fd, "w", encoding="utf-8", newline=newline) as f:
```

File: mdformat/_cli.py [hunk 4]
```diff
@@ -118,6 +118,20 @@ def validate_wrap_arg(value: str) -> Union[str, int]:
...
+        return "\n"
+    if value == "cr":
+        return "\r"
+    if value == "crlf":
+        return "\r\n"
+    if value == "system":
+        return None
+    raise ValueError("line-ending must be either a specifier or valid line ending.")
+
+
```
Comment (hukkin) [inline mdformat/_cli.py:134]:
```suggestion
```

File: mdformat/_cli.py [hunk 5]
```diff
@@ -149,6 +163,13 @@ def make_arg_parser(
         metavar="{keep,no,INTEGER}",
         help="paragraph word wrap mode (default: keep)",
     )
+    parser.add_argument(
+        "--line-ending",
+        default="system",
+        type=validate_line_ending_arg,
+        metavar="{system,lf,cr,crlf}",
+        help="output file line endings (default: system)",
+    )
```
Comment (hukkin) [inline mdformat/_cli.py:172]:
```suggestion
    parser.add_argument(
        "--end-of-line",
        default="lf",
        choices=("lf", "crlf"),
        help="output file line ending mode",
    )
```

File: mdformat/_cli.py [hunk 6]
```diff
@@ -101,7 +101,7 @@ def run(cli_args: Sequence[str]) -> int:  # noqa: C901
                 )
                 return 1
             if path:
-                atomic_write(path, formatted_str)
+                atomic_write(path, formatted_str, options.get("line_ending", None))
```
Comment (hukkin) [inline mdformat/_cli.py:104]:
```suggestion
                atomic_write(path, formatted_str, options["end_of_line"])
```

**Review by hukkin — CHANGES_REQUESTED** [review]
Thank you _very __very__ much_ for doing the research and coming up with an excellent fix. Much appreciated.

I've requested some changes to reduce the amount of configuration as I don't want to add any API surface unless a real world user with the need for it appears.

In addition to these changes we need to fix this line: https://github.com/executablebooks/mdformat/blob/ee13da698bcb24c1af05bf243b341845e4b60bdc/mdformat/_api.py#L62

We need to change it to 
```python
atomic_write(f, formatted_md, options.get("end_of_line", "lf"))
```

Also we need a simple test that ensures no `crlf`s are written to files by default. Running that in CI on Windows will validate our changes and prevent regressions. I can also add this test myself if you're not familiar with the test setup.

**Comment by hukkin** [comment]
I'm actually a bit tempted to not add the configuration option at all, and just always use a `\n` character always. This is what I thought the current behavior was until you proved otherwise :smile: 

I'm not sure if there's a good reason to prefer `\r\n` and if there's one person on earth who needs that, running in serial with an end of line fixer should solve that.

If it's later apparent that many people need `\r\n` then the configuration could be introduced.

What do you think @DragonCrafted87 ?

**Comment by DragonCrafted87** [comment]
yeah as while hardcoding it to purely `lf` would work in my case and I imagine most others I'm sure there's someone who wants all `crlf` line endings and like you said it will fail if there is an end of line fixer running in line with it in the same way I noticed it. 
I should be able to clean up the rest of the comments this evening though I may need a hand with the test as I tend to use python more for scripting than anything complicated enough to need a test suite

**Comment by hukkin** [comment]
:+1: 

For tests, please add these to `tests/test_cli.py`:
```python
def test_eol__lf(tmp_path):
    file_path = tmp_path / "test.md"
    file_path.write_bytes(b"Oi\r\n")
    assert run([str(file_path), "--end-of-line=lf"]) == 0
    assert file_path.read_bytes() == b"Oi\n"


def test_eol__crlf(tmp_path):
    file_path = tmp_path / "test.md"
    file_path.write_bytes(b"Oi\n")
    assert run([str(file_path), "--end-of-line=crlf"]) == 0
    assert file_path.read_bytes() == b"Oi\r\n"
```

and these to `tests/test_api.py`:
```python
def test_eol__lf(tmp_path):
    file_path = tmp_path / "test.md"
    file_path.write_bytes(b"Oi\r\n")
    mdformat.file(str(file_path), options={"end_of_line": "lf"})
    assert file_path.read_bytes() == b"Oi\n"


def test_eol__crlf(tmp_path):
    file_path = tmp_path / "test.md"
    file_path.write_bytes(b"Oi\n")
    mdformat.file(str(file_path), options={"end_of_line": "crlf"})
    assert file_path.read_bytes() == b"Oi\r\n"
```

**Comment by DragonCrafted87** [comment]
that push should have everything accounted for to just do the `crlf` and `lf` options. Thank you for the tests if I ever wind up doing a larger project in python I'll have to take a closer look at pytest as it looks like its nice and straight forward for a test system

**Comment by hukkin** [comment]
Thanks this is great! I'll try to have this released later today.

**Pull request merged**